Credit to https://github.com/harvitronix/neural-network-genetic-algorithm

Data set

In [ ]:
from keras.datasets import mnist
import numpy as np
from numpy import isneginf
from scipy.stats import iqr 
(X_train1, y_train1), (X_test1, y_test1) = mnist.load_data()


def compute_mag_phase(toBeTransfromed):
    dft = np.fft.fft2(toBeTransfromed)
    dft_shift = np.fft.fftshift(dft)
    mag = np.abs(dft_shift)
    ang = np.angle(dft_shift)
    return mag, ang

def reconstruct(mag,ang):
    combined = np.multiply(mag, np.exp(1j*ang))
    #combinedReal = np.real(combined)
    #combinedImag = np.imag(combined)
    fftx = np.fft.ifftshift(combined)
    ffty = np.fft.ifft2(fftx)
    imgCombined = np.abs(ffty)
    return imgCombined

def pre_process_mag(dataset):
    mag = np.zeros(dataset.shape)
    phase = np.zeros(dataset.shape)
    x= 0
    for i in range(0,dataset.shape[0]):
      mag[i,:,:],phase[i,:,:] = compute_mag_phase(dataset[i])
    return mag,phase

X_train_mag,X_train_phase = pre_process_mag(X_train1)
X_test_mag,X_test_phase = pre_process_mag(X_test1)

# flatten the magnitude
X_train_mag_flatten = X_train_mag.reshape(X_train_mag.shape[0], X_train_phase.shape[1] * X_train_mag.shape[2])
X_test_mag_flatten = X_test_mag.reshape(X_test_mag.shape[0], X_test_mag.shape[1] * X_test_mag.shape[2])

# flatten the phase
X_train_phase_flatten = X_train_phase.reshape(X_train_phase.shape[0], X_train_phase.shape[1] * X_train_phase.shape[2])
X_test_phase_flatten = X_test_phase.reshape(X_test_phase.shape[0], X_test_phase.shape[1] * X_test_phase.shape[2])


X_train_mag_flatten = X_train_mag_flatten.astype('float32')
X_test_mag_flatten = X_test_mag_flatten.astype('float32')

X_s = np.concatenate((X_train_mag_flatten,X_test_mag_flatten),axis =0)
IQR = iqr(X_s)
Median = np.median(X_s)

X_train_m_scaled = (X_train_mag_flatten - Median) / IQR
X_test_m_scaled = (X_test_mag_flatten - Median) / IQR



Training

In [ ]:
from keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from keras.utils.np_utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping

# Helper: Early stopping.
early_stopper = EarlyStopping(patience=5)

def get_mnist():
    """Retrieve the MNIST dataset and process the data."""
    # Set defaults.
    nb_classes = 10
    batch_size = 128
    input_shape = (784,)

    # Get the data.
    x_train = X_train_m_scaled
    x_test = X_test_m_scaled

    # convert class vectors to binary class matrices
    y_train = to_categorical(y_train1, nb_classes)
    y_test = to_categorical(y_test1, nb_classes)
    print('done mnist')
    return (nb_classes, batch_size, input_shape, x_train, x_test, y_train, y_test)

def compile_model(network, nb_classes, input_shape):
    """Compile a sequential model.
    Args:
        network (dict): the parameters of the network
    Returns:
        a compiled network.
    """
    # Get our network parameters.
    nb_layers = network['nb_layers']
    nb_neurons1 = network['nb_neurons1']
    nb_neurons2 = network['nb_neurons2']
    nb_neurons3 = network['nb_neurons3']
    nb_neurons4 = network['nb_neurons4']
    activation = network['activation']
    optimizer = network['optimizer']

    model = Sequential()

    # Add each layer.
    for i in range(nb_layers):

        # Need input shape for first layer.
        if i == 0:
            model.add(Dense(nb_neurons1, activation=activation, input_shape=input_shape))
        elif i == 1:
            model.add(Dense(nb_neurons2, activation=activation))
        elif i == 2:
            model.add(Dense(nb_neurons3, activation=activation))
        elif i == 3:
            model.add(Dense(nb_neurons4, activation=activation))

        model.add(Dropout(0.2))  # hard-coded dropout

    # Output layer.
    model.add(Dense(nb_classes, activation='softmax'))

    model.compile(loss='categorical_crossentropy', optimizer=optimizer,
                  metrics=['accuracy'])
    print('compiled model')
    return model

def train_and_score(network, dataset):
    """Train the model, return test loss.
    Args:
        network (dict): the parameters of the network
        dataset (str): Dataset to use for training/evaluating
    """
    logging.info("***train_and_score***")
    if dataset == 'cifar10':
        nb_classes, batch_size, input_shape, x_train, \
            x_test, y_train, y_test = get_cifar10()
    elif dataset == 'mnist':
        logging.info("***mnist***")
        nb_classes, batch_size, input_shape, x_train, \
            x_test, y_train, y_test = get_mnist()

    model = compile_model(network, nb_classes, input_shape)
    logging.info("***compiled_model***")
    model.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=100,  # using early stopping, so no real limit
              verbose=0,
              validation_split=0.2,
              callbacks=[early_stopper])
    #validation_split=0.2, epochs=100, batch_size=2000,
    #validation_data=(x_test, y_test),
    score = model.evaluate(x_test, y_test, verbose=0)
    logging.info("***trainedd***")
    print(score[1])
    return score[1]  # 1 is accuracy. 0 is loss.

Network

In [ ]:
import random
import logging
#from train import train_and_score

class Network():
    """Represent a network and let us operate on it.
    Currently only works for an MLP.
    """

    def __init__(self, nn_param_choices=None):
        """Initialize our network.
        Args:
            nn_param_choices (dict): Parameters for the network, includes:
                nb_neurons (list): [64, 128, 256]
                nb_layers (list): [1, 2, 3, 4]
                activation (list): ['relu', 'elu']
                optimizer (list): ['rmsprop', 'adam']
        """
        self.accuracy = 0.
        self.nn_param_choices = nn_param_choices
        self.network = {}  # (dic): represents MLP network parameters

    def create_random(self):
        i = True 
        """Create a random network."""
        for key in self.nn_param_choices:
            if key == 'nb_neurons2':
              while (True):
                logging.info("stuck 1")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons1']:
                  break
            elif key == 'nb_neurons3':
              while (True):
                logging.info("stuck 2")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons2']:
                  break
            elif key == 'nb_neurons4':
              while (True):
                logging.info("stuck 3")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons3']:
                  break
            else:
              self.network[key] = random.choice(self.nn_param_choices[key])



    def create_set(self, network):
        """Set network properties.
        Args:
            network (dict): The network parameters
        """
        self.network = network

    def train(self, dataset):
        """Train the network and record the accuracy.
        Args:
            dataset (str): Name of dataset to use.
        """
        logging.info("***train***")
        if self.accuracy == 0.:
            self.accuracy = train_and_score(self.network, dataset)

    def print_network(self):
        """Print out a network."""
        logging.info(self.network)
        logging.info("Network accuracy: %.2f%%" % (self.accuracy * 100))

In [ ]:
from functools import reduce
from operator import add
import random


class Optimizer():
    """Class that implements genetic algorithm for MLP optimization."""

    def __init__(self, nn_param_choices, retain=0.4,
                 random_select=0.1, mutate_chance=0.3):
        """Create an optimizer.
        Args:
            nn_param_choices (dict): Possible network paremters
            retain (float): Percentage of population to retain after
                each generation
            random_select (float): Probability of a rejected network
                remaining in the population
            mutate_chance (float): Probability a network will be
                randomly mutated
        """
        self.mutate_chance = mutate_chance
        self.random_select = random_select
        self.retain = retain
        self.nn_param_choices = nn_param_choices

    def create_population(self, count):
        """Create a population of random networks.
        Args:
            count (int): Number of networks to generate, aka the
                size of the population
        Returns:
            (list): Population of network objects
        """
        pop = []
        for _ in range(0, count):
            # Create a random network.
            network = Network(self.nn_param_choices)
            logging.info("count")
            network.create_random()

            # Add the network to our population.
            pop.append(network)

        return pop

    @staticmethod
    def fitness(network):
        """Return the accuracy, which is our fitness function."""
        return network.accuracy

    def grade(self, pop):
        """Find average fitness for a population.
        Args:
            pop (list): The population of networks
        Returns:
            (float): The average accuracy of the population
        """
        summed = reduce(add, (self.fitness(network) for network in pop))
        return summed / float((len(pop)))

    def breed(self, mother, father):
        """Make two children as parts of their parents.
        Args:
            mother (dict): Network parameters
            father (dict): Network parameters
        Returns:
            (list): Two network objects
        """
        children = []
        for _ in range(2):

            child = {}

            #  need to do if they have equal layers
            # Loop through the parameters and pick params for the kid.
            for param in self.nn_param_choices:
                  if (param == 'nb_neurons2') :
                    while (True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons1']:
                            logging.info("***broke1***")
                            break
                  elif param == 'nb_neurons3':
                    while ( True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons2']:
                            break
                            logging.info("***broke2***")
                  elif param == 'nb_neurons4':
                    while (True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons3']:
                            break
                            logging.info("***broke3***")
                  else:
                    child[param] = random.choice([mother.network[param], father.network[param]])

            # Now create a network object.
            network = Network(self.nn_param_choices)
            network.create_set(child)

            # Randomly mutate some of the children.
            if self.mutate_chance > random.random():
                network = self.mutate(network)

            children.append(network)

        return children

    def mutate(self, network):
        """Randomly mutate one part of the network.
        Args:
            network (dict): The network parameters to mutate
        Returns:
            (Network): A randomly mutated network object
        """
        # Choose a random key.

 
        mutation = random.choice(list(self.nn_param_choices.keys()))
        
        #mutate param within condition
        if mutation == 'nb_neurons1':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons2']):
                    break
        elif mutation == 'nb_neurons2':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons3']) and (network.network[mutation] < network.network['nb_neurons1']) :
                    break
        elif mutation == 'nb_neurons3':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons4'])  and (network.network[mutation] < network.network['nb_neurons2']):
                    break
        elif mutation == 'nb_neurons4':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] < network.network['nb_neurons3']):
                    break
        else:
          # Mutate one of the params.
          network.network[mutation] = random.choice(self.nn_param_choices[mutation])
        return network

    def evolve(self, pop):
        """Evolve a population of networks.
        Args:
            pop (list): A list of network parameters
        Returns:
            (list): The evolved population of networks
        """
        # Get scores for each network.
        graded = [(self.fitness(network), network) for network in pop]

        # Sort on the scores.
        graded = [x[1] for x in sorted(graded, key=lambda x: x[0], reverse=True)]

        # Get the number we want to keep for the next gen.
        retain_length = int(len(graded)*self.retain)

        # The parents are every network we want to keep.
        parents = graded[:retain_length]

        # For those we aren't keeping, randomly keep some anyway.
        for individual in graded[retain_length:]:
            if self.random_select > random.random():
                parents.append(individual)

        # Now find out how many spots we have left to fill.
        parents_length = len(parents)
        desired_length = len(pop) - parents_length
        children = []

        # Add children, which are bred from two remaining networks.
        while len(children) < desired_length:

            # Get a random mom and dad.
            print(parents_length-1)
            male = random.randint(0, parents_length-1)

            female = random.randint(0, parents_length-1)

            # Assuming they aren't the same network...
            if male != female:
                male = parents[male]
                female = parents[female]

                # Breed them.
                babies = self.breed(male, female)

                # Add the children one at a time.
                for baby in babies:
                    # Don't grow larger than desired length.
                    if len(children) < desired_length:
                        children.append(baby)

        parents.extend(children)

        return parents

In [ ]:
import logging
from tqdm import tqdm

# Setup logging.
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filename='log.txt'
)

def train_networks(networks, dataset):
    """Train each network.
    Args:
        networks (list): Current population of networks
        dataset (str): Dataset to use for training/evaluating
    """

    pbar = tqdm(total=len(networks))
    logging.info("***a***")
    for network in networks:
        network.train(dataset)
        pbar.update(1)
        logging.info("***b***")
    pbar.close()

def get_average_accuracy(networks):
    """Get the average accuracy for a group of networks.
    Args:
        networks (list): List of networks
    Returns:
        float: The average accuracy of a population of networks.
    """
    total_accuracy = 0
    for network in networks:
        total_accuracy += network.accuracy

    return total_accuracy / len(networks)

def generate(generations, population, nn_param_choices, dataset):
    """Generate a network with the genetic algorithm.
    Args:
        generations (int): Number of times to evole the population
        population (int): Number of networks in each generation
        nn_param_choices (dict): Parameter choices for networks
        dataset (str): Dataset to use for training/evaluating
    """
    optimizer = Optimizer(nn_param_choices)
    logging.info("***Optimizer***")
    networks = optimizer.create_population(population)

    # Evolve the generation.
    for i in range(generations):
        logging.info("***Doing generation %d of %d***" %
                     (i + 1, generations))

        # Train and get accuracy for networks.
        train_networks(networks, dataset)
        logging.info("***Trained networks***")
        # Get the average accuracy for this generation.
        average_accuracy = get_average_accuracy(networks)

        # Print out the average accuracy each generation.
        logging.info("Generation average: %.2f%%" % (average_accuracy * 100))
        logging.info('-'*80)

        # Evolve, except on the last iteration.
        if i != generations - 1:
            # Do the evolution.
            networks = optimizer.evolve(networks)

    # Sort our final population.
    networks = sorted(networks, key=lambda x: x.accuracy, reverse=True)

    # Print out the top 5 networks.
    print_networks(networks[:5])

def print_networks(networks):
    """Print a list of networks.
    Args:
        networks (list): The population of networks
    """
    logging.info('-'*80)
    for network in networks:
        network.print_network()

def main():
    """Evolve a network."""
    generations = 7  # Number of times to evole the population.
    population = 30  # Number of networks in each generation.
    dataset = 'mnist'
    print(2)

    nn_param_choices = {
        #'nb_neurons': [10, 60, 110, 160, 210, 260,310,360,410,460,510],
        'nb_neurons1': range(14,780),
        'nb_neurons2':  range(13,779),
        'nb_neurons3': range(12,778),
        'nb_neurons4': range(11,777),
        'nb_layers': [1, 2, 3, 4],
        'activation': ['relu', 'elu', 'tanh', 'sigmoid'],
        'optimizer': ['rmsprop', 'adam', 'sgd', 'adagrad',
                      'adadelta', 'adamax', 'nadam'],
    }
    print(3)
    logging.info("***Evolving %d generations with population %d***" %
                 (generations, population))
    print(4)
    generate(generations, population, nn_param_choices, dataset)
    print('done')

if __name__ == '__main__':
    main()

  0%|          | 0/30 [00:00<?, ?it/s]

2
3
4
done mnist
compiled model


  3%|▎         | 1/30 [01:29<43:17, 89.56s/it]

0.9093999862670898
done mnist
compiled model


  7%|▋         | 2/30 [01:42<31:00, 66.45s/it]

0.9333000183105469
done mnist
compiled model


 10%|█         | 3/30 [02:47<29:49, 66.28s/it]

0.8837000131607056
done mnist
compiled model


 13%|█▎        | 4/30 [03:05<22:20, 51.55s/it]

0.9355999827384949
done mnist
compiled model


 17%|█▋        | 5/30 [03:36<18:53, 45.35s/it]

0.9383000135421753
done mnist
compiled model


 20%|██        | 6/30 [04:16<17:30, 43.78s/it]

0.9423999786376953
done mnist
compiled model


 23%|██▎       | 7/30 [05:28<20:03, 52.31s/it]

0.9312000274658203
done mnist
compiled model


 27%|██▋       | 8/30 [05:37<14:25, 39.34s/it]

0.9244999885559082
done mnist
compiled model


 30%|███       | 9/30 [06:45<16:46, 47.93s/it]

0.8619999885559082
done mnist
compiled model


 33%|███▎      | 10/30 [08:06<19:17, 57.86s/it]

0.8270000219345093
done mnist
compiled model


 37%|███▋      | 11/30 [08:29<15:00, 47.38s/it]

0.9215999841690063
done mnist
compiled model


 40%|████      | 12/30 [08:39<10:53, 36.33s/it]

0.9294000267982483
done mnist
compiled model


 43%|████▎     | 13/30 [10:03<14:17, 50.44s/it]

0.8363000154495239
done mnist
compiled model


 47%|████▋     | 14/30 [10:54<13:32, 50.81s/it]

0.8267999887466431
done mnist
compiled model


 50%|█████     | 15/30 [12:12<14:42, 58.82s/it]

0.9020000100135803
done mnist
compiled model


 53%|█████▎    | 16/30 [13:27<14:50, 63.63s/it]

0.8999999761581421
done mnist
compiled model


 57%|█████▋    | 17/30 [13:44<10:44, 49.57s/it]

0.9423999786376953
done mnist
compiled model


 60%|██████    | 18/30 [14:01<07:58, 39.89s/it]

0.9347000122070312
done mnist
compiled model


 63%|██████▎   | 19/30 [14:29<06:39, 36.30s/it]

0.9437999725341797
done mnist
compiled model


 67%|██████▋   | 20/30 [15:40<07:48, 46.89s/it]

0.7016000151634216
done mnist
compiled model


 70%|███████   | 21/30 [16:11<06:16, 41.85s/it]

0.9390000104904175
done mnist
compiled model


 73%|███████▎  | 22/30 [17:28<07:01, 52.64s/it]

0.8964999914169312
done mnist
compiled model


 77%|███████▋  | 23/30 [18:39<06:45, 57.98s/it]

0.9300000071525574
done mnist
compiled model


 80%|████████  | 24/30 [19:54<06:18, 63.13s/it]

0.8920000195503235
done mnist
compiled model


 83%|████████▎ | 25/30 [20:38<04:46, 57.36s/it]

0.9465000033378601
done mnist
compiled model


 87%|████████▋ | 26/30 [20:56<03:01, 45.50s/it]

0.9352999925613403
done mnist
compiled model


 90%|█████████ | 27/30 [21:17<01:54, 38.27s/it]

0.9329000115394592
done mnist
compiled model


 93%|█████████▎| 28/30 [23:03<01:57, 58.70s/it]

0.7753999829292297
done mnist
compiled model


 97%|█████████▋| 29/30 [24:12<01:01, 61.54s/it]

0.8544999957084656
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.4239000082015991
14
14
14
14
14
14
14
14
done mnist
compiled model


 53%|█████▎    | 16/30 [00:14<00:12,  1.09it/s]

0.9222000241279602
done mnist
compiled model


 57%|█████▋    | 17/30 [00:36<01:34,  7.26s/it]

0.9275000095367432
done mnist
compiled model


 60%|██████    | 18/30 [00:52<01:58,  9.86s/it]

0.9387999773025513
done mnist
compiled model


 63%|██████▎   | 19/30 [02:10<05:31, 30.12s/it]

0.8826000094413757
done mnist
compiled model


 67%|██████▋   | 20/30 [02:21<04:06, 24.63s/it]

0.9419999718666077
done mnist
compiled model


 70%|███████   | 21/30 [02:47<03:44, 25.00s/it]

0.9366999864578247
done mnist
compiled model


 73%|███████▎  | 22/30 [04:08<05:33, 41.73s/it]

0.828499972820282
done mnist
compiled model


 77%|███████▋  | 23/30 [05:23<06:01, 51.57s/it]

0.9016000032424927
done mnist
compiled model


 80%|████████  | 24/30 [06:15<05:11, 51.97s/it]

0.9437999725341797
done mnist
compiled model


 83%|████████▎ | 25/30 [07:01<04:10, 50.01s/it]

0.9398000240325928
done mnist
compiled model


 87%|████████▋ | 26/30 [07:26<02:50, 42.50s/it]

0.9200000166893005
done mnist
compiled model


 90%|█████████ | 27/30 [08:45<02:40, 53.48s/it]

0.8758000135421753
done mnist
compiled model


 93%|█████████▎| 28/30 [08:56<01:21, 40.69s/it]

0.9291999936103821
done mnist
compiled model


 97%|█████████▋| 29/30 [09:12<00:33, 33.37s/it]

0.9283000230789185
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.9456999897956848
14
14
14
14
14
14
14
14
done mnist
compiled model


 53%|█████▎    | 16/30 [00:14<00:12,  1.14it/s]

0.9309999942779541
done mnist
compiled model


 57%|█████▋    | 17/30 [00:32<01:19,  6.09s/it]

0.9413999915122986
done mnist
compiled model


 60%|██████    | 18/30 [00:46<01:43,  8.66s/it]

0.944100022315979
done mnist
compiled model


 63%|██████▎   | 19/30 [01:09<02:19, 12.68s/it]

0.9430999755859375
done mnist
compiled model


 67%|██████▋   | 20/30 [01:25<02:18, 13.89s/it]

0.9401999711990356
done mnist
compiled model


 70%|███████   | 21/30 [02:12<03:33, 23.70s/it]

0.9488999843597412
done mnist
compiled model


 73%|███████▎  | 22/30 [02:54<03:54, 29.35s/it]

0.9391000270843506
done mnist
compiled model


 77%|███████▋  | 23/30 [03:29<03:36, 30.89s/it]

0.9445000290870667
done mnist
compiled model


 80%|████████  | 24/30 [03:46<02:39, 26.66s/it]

0.9369000196456909
done mnist
compiled model


 83%|████████▎ | 25/30 [04:01<01:56, 23.25s/it]

0.9390000104904175
done mnist
compiled model


 87%|████████▋ | 26/30 [05:18<02:37, 39.41s/it]

0.8294000029563904
done mnist
compiled model


 90%|█████████ | 27/30 [06:15<02:13, 44.57s/it]

0.9286999702453613
done mnist
compiled model


 93%|█████████▎| 28/30 [06:31<01:12, 36.11s/it]

0.9311000108718872
done mnist
compiled model


 97%|█████████▋| 29/30 [06:51<00:31, 31.29s/it]

0.9431999921798706
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.940500020980835
12
12
12
12
12
12
12
12
12
done mnist
compiled model


 47%|████▋     | 14/30 [00:23<00:26,  1.66s/it]

0.9351000189781189
done mnist
compiled model


 50%|█████     | 15/30 [01:01<03:09, 12.61s/it]

0.9455000162124634
done mnist
compiled model


 53%|█████▎    | 16/30 [01:41<04:52, 20.91s/it]

0.9354000091552734
done mnist
compiled model


 57%|█████▋    | 17/30 [02:25<06:01, 27.83s/it]

0.9337999820709229
done mnist
compiled model


 60%|██████    | 18/30 [03:11<06:37, 33.13s/it]

0.9447000026702881
done mnist
compiled model


 63%|██████▎   | 19/30 [03:36<05:39, 30.89s/it]

0.9416000247001648
done mnist
compiled model


 67%|██████▋   | 20/30 [04:10<05:15, 31.59s/it]

0.9456999897956848
done mnist
compiled model


 70%|███████   | 21/30 [04:46<04:56, 32.95s/it]

0.9445000290870667
done mnist
compiled model


 73%|███████▎  | 22/30 [05:03<03:46, 28.33s/it]

0.9419000148773193
done mnist
compiled model


 77%|███████▋  | 23/30 [06:22<05:04, 43.49s/it]

0.8672000169754028
done mnist
compiled model


 80%|████████  | 24/30 [06:40<03:34, 35.82s/it]

0.9395999908447266
done mnist
compiled model


 83%|████████▎ | 25/30 [06:55<02:27, 29.57s/it]

0.9312999844551086
done mnist
compiled model


 87%|████████▋ | 26/30 [07:11<01:42, 25.66s/it]

0.9397000074386597
done mnist
compiled model


 90%|█████████ | 27/30 [07:26<01:06, 22.26s/it]

0.9473999738693237
done mnist
compiled model


 93%|█████████▎| 28/30 [08:02<00:52, 26.32s/it]

0.945900022983551
done mnist
compiled model


 97%|█████████▋| 29/30 [08:41<00:30, 30.26s/it]

0.9469000101089478
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.944599986076355
14
14
14
14
14
14
14
14
done mnist
compiled model


 53%|█████▎    | 16/30 [00:43<00:37,  2.71s/it]

0.944599986076355
done mnist
compiled model


 57%|█████▋    | 17/30 [01:35<03:48, 17.61s/it]

0.9447000026702881
done mnist
compiled model


 60%|██████    | 18/30 [02:26<05:29, 27.48s/it]

0.945900022983551
done mnist
compiled model


 63%|██████▎   | 19/30 [03:11<06:02, 32.93s/it]

0.9463000297546387
done mnist
compiled model


 67%|██████▋   | 20/30 [03:53<05:55, 35.52s/it]

0.9480000138282776
done mnist
compiled model


 70%|███████   | 21/30 [04:14<04:40, 31.12s/it]

0.9401999711990356
done mnist
compiled model


 73%|███████▎  | 22/30 [05:00<04:45, 35.75s/it]

0.9470999836921692
done mnist
compiled model


 77%|███████▋  | 23/30 [05:16<03:26, 29.57s/it]

0.9441999793052673
done mnist
compiled model


 80%|████████  | 24/30 [05:46<02:58, 29.81s/it]

0.942799985408783
done mnist
compiled model


 83%|████████▎ | 25/30 [06:22<02:38, 31.70s/it]

0.9460999965667725
done mnist
compiled model


 87%|████████▋ | 26/30 [06:55<02:08, 32.03s/it]

0.9427000284194946
done mnist
compiled model


 90%|█████████ | 27/30 [07:30<01:38, 32.94s/it]

0.9409999847412109
done mnist
compiled model


 93%|█████████▎| 28/30 [08:17<01:14, 37.16s/it]

0.947700023651123
done mnist
compiled model


 97%|█████████▋| 29/30 [09:00<00:38, 39.00s/it]

0.9438999891281128
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.9474999904632568
13
13
13
13
13
13
13
13
done mnist
compiled model


 50%|█████     | 15/30 [00:31<00:31,  2.09s/it]

0.9413999915122986
done mnist
compiled model


 53%|█████▎    | 16/30 [01:20<03:48, 16.29s/it]

0.946399986743927
done mnist
compiled model


 57%|█████▋    | 17/30 [01:48<04:17, 19.77s/it]

0.9420999884605408
done mnist
compiled model


 60%|██████    | 18/30 [02:34<05:30, 27.53s/it]

0.9476000070571899
done mnist
compiled model


 63%|██████▎   | 19/30 [03:12<05:38, 30.75s/it]

0.9420999884605408
done mnist
compiled model


 67%|██████▋   | 20/30 [03:59<05:56, 35.66s/it]

0.9452999830245972
done mnist
compiled model


 70%|███████   | 21/30 [04:40<05:36, 37.35s/it]

0.9435999989509583
done mnist
compiled model


 73%|███████▎  | 22/30 [05:46<06:07, 45.95s/it]

0.9453999996185303
done mnist
compiled model


 77%|███████▋  | 23/30 [06:40<05:38, 48.33s/it]

0.9470999836921692
done mnist
compiled model


 80%|████████  | 24/30 [07:11<04:17, 42.92s/it]

0.9448999762535095
done mnist
compiled model


 83%|████████▎ | 25/30 [07:58<03:41, 44.26s/it]

0.9473000168800354
done mnist
compiled model


 87%|████████▋ | 26/30 [08:32<02:44, 41.12s/it]

0.9441999793052673
done mnist
compiled model


 90%|█████████ | 27/30 [09:27<02:15, 45.23s/it]

0.9506000280380249
done mnist
compiled model


 93%|█████████▎| 28/30 [10:06<01:27, 43.55s/it]

0.949400007724762
done mnist
compiled model


 97%|█████████▋| 29/30 [11:28<00:55, 55.03s/it]

0.8830000162124634
done mnist
compiled model


  0%|          | 0/30 [00:00<?, ?it/s]

0.9437000155448914
14
14
14
14
14
14
14
14
done mnist
compiled model


 53%|█████▎    | 16/30 [00:47<00:41,  2.99s/it]

0.9434999823570251
done mnist
compiled model


 57%|█████▋    | 17/30 [01:35<03:33, 16.40s/it]

0.947700023651123
done mnist
compiled model


 60%|██████    | 18/30 [02:08<04:16, 21.37s/it]

0.9452000260353088
done mnist
compiled model


 63%|██████▎   | 19/30 [02:46<04:50, 26.41s/it]

0.9438999891281128
done mnist
compiled model


 67%|██████▋   | 20/30 [03:21<04:48, 28.85s/it]

0.9451000094413757
done mnist
compiled model


 70%|███████   | 21/30 [04:01<04:50, 32.29s/it]

0.9412000179290771
done mnist
compiled model


 73%|███████▎  | 22/30 [04:39<04:31, 33.89s/it]

0.9433000087738037
done mnist
compiled model


 77%|███████▋  | 23/30 [05:27<04:27, 38.23s/it]

0.9431999921798706
done mnist
compiled model


 80%|████████  | 24/30 [06:03<03:45, 37.51s/it]

0.9449999928474426
done mnist
compiled model


 83%|████████▎ | 25/30 [06:55<03:28, 41.77s/it]

0.9452000260353088
done mnist
compiled model


 87%|████████▋ | 26/30 [08:09<03:26, 51.61s/it]

0.9057000279426575
done mnist
compiled model


 90%|█████████ | 27/30 [08:53<02:28, 49.43s/it]

0.946399986743927
done mnist
compiled model


 93%|█████████▎| 28/30 [09:43<01:38, 49.36s/it]

0.9491999745368958
done mnist
compiled model


 97%|█████████▋| 29/30 [10:14<00:43, 43.87s/it]

0.9433000087738037
done mnist
compiled model


100%|██████████| 30/30 [11:02<00:00, 22.09s/it]

0.9445000290870667
done


{'nb_neurons1': 503, 'nb_neurons2': 500, 'nb_neurons3': 66, 'nb_neurons4': 48, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'sgd'}
Network accuracy: 95.06%

{'nb_neurons1': 503, 'nb_neurons2': 500, 'nb_neurons3': 66, 'nb_neurons4': 48, 'nb_layers': 3, 'activation': 'elu', 'optimizer': 'sgd'}
Network accuracy: 94.94%

{'nb_neurons1': 688, 'nb_neurons2': 310, 'nb_neurons3': 96, 'nb_neurons4': 29, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'sgd'}
Network accuracy: 94.92%

{'nb_neurons1': 407, 'nb_neurons2': 310, 'nb_neurons3': 182, 'nb_neurons4': 130, 'nb_layers': 4, 'activation': 'elu', 'optimizer': 'sgd'}
Network accuracy: 94.89%

{'nb_neurons1': 639, 'nb_neurons2': 500, 'nb_neurons3': 66, 'nb_neurons4': 48, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'sgd'}
Network accuracy: 94.80%

Potential to add starting points.